common/src/main/java/org/keycloak/common/DeviceInfo.java

```diff
@@ -0,0 +1,95 @@
+package org.keycloak.common;
+
+public class DeviceInfo {
+    public static String ID = "DEVICE_INFO";
+    private String device;
+    private String browser;
+    private String browserVersion;
+    private String os;
+    private String osVersion;
+    private String userAgent;
+
+    public DeviceInfo() {
+    }
+
+    public DeviceInfo(String device, String browser, String browserVersion, String os, String osVersion, String userAgent) {
+        this.device = device;
+        this.browser = browser;
+        this.browserVersion = browserVersion;
+        this.os = os;
+        this.osVersion = osVersion;
+        this.userAgent = userAgent;
+    }
+
+    public static DeviceInfo create(String deviceInfo) {
+        if (deviceInfo != null && deviceInfo.matches(".*:.*:.*:.*:.*:.*")) {
+            String[] info = deviceInfo.split(":");
+            return new DeviceInfo(info[0], info[1], info[2], info[3], info[4], info[5]);
+        } else
+            return new DeviceInfo("", "", "", "", "", "");
+    }
+
+    public String getDevice() {
+        return device;
+    }
+
+    public void setDevice(String device) {
+        this.device = device;
+    }
+
+    public String getBrowser() {
+        return browser;
+    }
+
+    public void setBrowser(String browser) {
+        this.browser = browser;
+    }
+
+    public String getBrowserVersion() {
+        return browserVersion;
+    }
+
+    public void setBrowserVersion(String browserVersion) {
+        this.browserVersion = browserVersion;
+    }
+
+    public String getOs() {
+        return os;
+    }
+
+    public void setOs(String os) {
+        this.os = os;
+    }
+
+    public String getOsVersion() {
+        return osVersion;
+    }
+
+    public void setOsVersion(String osVersion) {
+        this.osVersion = osVersion;
+    }
+
+    public String getUserAgent() {
+        return userAgent;
+    }
+
+    public void setUserAgent(String userAgent) {
+        this.userAgent = userAgent;
+    }
+
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
```

Comment:
Or other alternative can be to use other very strange delimiter like _@@_ when there is no chance of having the delimiter in the values and hence no escaping needed.

Reply:
Regardless of the actual separator shape, it needs to be escaped, since it is coming from HTTP headers, i.e. potentially malicious source. In this case any occurrence of `@@` would need escaping.

Reply:
@hmlnarik Good point regarding escaping. With regards to this, I don;t have any preference whether to use ":" or something like "@@" . Can be fine to stick with ":" if it's easier.

Reply:
Incidentally worked on another issue that also needs multi-field string serialization, and developed a tool for fast string (de)serialization, maybe it is worth reusing: https://github.com/keycloak/keycloak/pull/6139/files#diff-de17a5622aadad5857ba5697419c1e0b